src/Exceptionless.Core/Services/StackService.cs

```diff
@@ -0,0 +1,97 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Threading;
+using System.Threading.Tasks;
+using Exceptionless.Core.Extensions;
+using Exceptionless.Core.Repositories;
+using Exceptionless.DateTimeExtensions;
+using Foundatio.Caching;
+using Foundatio.Utility;
+using Microsoft.Extensions.Logging;
+
+namespace Exceptionless.Core.Services {
+    public class StackService {
+        private readonly ILogger<UsageService> _logger;
+        private readonly IStackRepository _stackRepository;
+        private readonly ICacheClient _cache;
+        private readonly TimeSpan _expireTimeout = TimeSpan.FromHours(12);
+
+        public StackService(IStackRepository stackRepository, ICacheClient cache, ILoggerFactory loggerFactory = null) {
+            _stackRepository = stackRepository;
+            _cache = cache;
+            _logger = loggerFactory.CreateLogger<UsageService>();
+        }
+
+        public async Task IncrementStackUsageAsync(string organizationId, string projectId, string stackId, DateTime minOccurrenceDateUtc, DateTime maxOccurrenceDateUtc, int count) {
+            if (String.IsNullOrEmpty(organizationId))
+                throw new ArgumentNullException(nameof(organizationId));
+            if (String.IsNullOrEmpty(projectId))
+                throw new ArgumentNullException(nameof(projectId));
+            if (String.IsNullOrEmpty(stackId))
+                throw new ArgumentNullException(nameof(stackId));
+            if (count == 0)
+                return;
+
+            var tasks = new List<Task>(4);
+
+            string occurenceCountCacheKey = GetStackOccurrenceCountCacheKey(organizationId, projectId, stackId),
+                occurrenceMinDateCacheKey = GetStackOccurrenceMinDateCacheKey(organizationId, projectId, stackId),
+                occurrenceMaxDateCacheKey = GetStackOccurrenceMaxDateCacheKey(organizationId, projectId, stackId),
+                occurrenceSetCacheKey = GetStackOccurrenceSetCacheKey();
+
+            var cachedOccurrenceMinDateUtc = await _cache.GetAsync<DateTime>(occurrenceMinDateCacheKey).AnyContext();
+            if (!cachedOccurrenceMinDateUtc.HasValue || cachedOccurrenceMinDateUtc.Value.IsAfter(minOccurrenceDateUtc))
+                tasks.Add(_cache.SetAsync(occurrenceMinDateCacheKey, minOccurrenceDateUtc, _expireTimeout));
+
+            var cachedOccurrenceMaxDateUtc = await _cache.GetAsync<DateTime>(occurrenceMaxDateCacheKey).AnyContext();
+            if (!cachedOccurrenceMaxDateUtc.HasValue || cachedOccurrenceMaxDateUtc.Value.IsBefore(maxOccurrenceDateUtc))
+                tasks.Add(_cache.SetAsync(occurrenceMaxDateCacheKey, maxOccurrenceDateUtc, _expireTimeout));
```

Comment:
This should be using the SetIfLowerAsync and using the long epoch value so that we don't have to do a get and a set which could also be a race condition.